File: src/main/java/com/radar/prospection/scraper/ScraperCoordinator.java
package com.radar.prospection.scraper;

import com.radar.prospection.claude.MissionAnalysisService;
import com.radar.prospection.config.RadarProperties;
import com.radar.prospection.domain.Mission;
import com.radar.prospection.repository.MissionRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

@Slf4j
@Service
@RequiredArgsConstructor
public class ScraperCoordinator {

    private final List<MissionScraper> scrapers;
    private final MissionRepository missionRepository;
    private final MissionAnalysisService analysisService;
    private final RadarProperties properties;

    public ScrapingResult runAll() {
        List<String> keywords = properties.getScraping().getKeywords();
        AtomicInteger newCount = new AtomicInteger(0);
        AtomicInteger duplicateCount = new AtomicInteger(0);
        AtomicInteger analyzedCount = new AtomicInteger(0);
        List<String> errors = new ArrayList<>();

        log.info("=== RADAR SCAN démarré — {} scrapers, {} keywords ===",
            scrapers.size(), keywords.size());

        for (MissionScraper scraper : scrapers) {
            log.info("--- Scraper: {} ---", scraper.getSource());
            try {
                List<Mission> found = scraper.scrape(keywords);
                log.info("[{}] {} missions brutes trouvées", scraper.getSource(), found.size());

                for (Mission mission : found) {
                    try {
                        if (missionRepository.existsBySourceAndExternalId(
                                mission.getSource(), mission.getExternalId())) {
                            duplicateCount.incrementAndGet();
                            continue;
                        }

                        // Analyse Claude avant sauvegarde
                        Mission analyzed = analysisService.analyze(mission);
                        missionRepository.save(analyzed);
                        newCount.incrementAndGet();
                        analyzedCount.incrementAndGet();

                    } catch (Exception e) {
                        log.warn("[{}] Erreur sauvegarde mission '{}': {}",
                            scraper.getSource(), mission.getTitle(), e.getMessage());
                        errors.add(scraper.getSource() + ": " + e.getMessage());
                    }
                }

            } catch (Exception e) {
                log.error("[{}] Scraper en erreur: {}", scraper.getSource(), e.getMessage());
                errors.add(scraper.getSource() + " FATAL: " + e.getMessage());
            }
        }

        log.info("=== RADAR SCAN terminé — {} nouvelles, {} doublons, {} erreurs ===",
            newCount.get(), duplicateCount.get(), errors.size());

        return new ScrapingResult(newCount.get(), duplicateCount.get(), analyzedCount.get(), errors);
    }

    public record ScrapingResult(
        int newMissions,
        int duplicates,
        int analyzed,
        List<String> errors
    ) {}
}

File: backend/src/main/kotlin/com/radar/prospection/scraper/ScraperCoordinator.kt
package com.radar.prospection.scraper

import com.radar.prospection.claude.MissionAnalysisService
import com.radar.prospection.config.RadarProperties
import com.radar.prospection.notification.NotificationService
import com.radar.prospection.repository.MissionRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class ScraperCoordinator(
    private val scrapers: List<MissionScraper>,
    private val missionRepository: MissionRepository,
    private val analysisService: MissionAnalysisService,
    private val properties: RadarProperties,
    private val notificationService: NotificationService? = null
) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun runAll(): ScrapingResult {
        val keywords = properties.scraping.keywords
        var newCount = 0
        var duplicateCount = 0
        var analyzedCount = 0
        val errors = mutableListOf<String>()

        log.info("=== RADAR SCAN démarré — {} scrapers, {} keywords ===", scrapers.size, keywords.size)

        for (scraper in scrapers) {
            log.info("--- Scraper: {} ---", scraper.getSource())
            try {
                val found = scraper.scrape(keywords)
                log.info("[{}] {} missions brutes trouvées", scraper.getSource(), found.size)

                for (mission in found) {
                    try {
                        if (missionRepository.existsBySourceAndExternalId(mission.source!!, mission.externalId!!)) {
                            duplicateCount++
                            continue
                        }
                        val analyzed = analysisService.analyze(mission)
                        val saved = missionRepository.save(analyzed)
                        notificationService?.notifyHighScoreMission(saved)
                        newCount++
                        analyzedCount++
                    } catch (e: Exception) {
                        log.warn("[{}] Erreur sauvegarde mission '{}': {}", scraper.getSource(), mission.title, e.message)
                        errors.add("${scraper.getSource()}: ${e.message}")
                    }
                }
            } catch (e: Exception) {
                log.error("[{}] Scraper en erreur: {}", scraper.getSource(), e.message)
                errors.add("${scraper.getSource()} FATAL: ${e.message}")
            }
        }

        log.info("=== RADAR SCAN terminé — {} nouvelles, {} doublons, {} erreurs ===", newCount, duplicateCount, errors.size)
        return ScrapingResult(newCount, duplicateCount, analyzedCount, errors)
    }

    data class ScrapingResult(
        val newMissions: Int,
        val duplicates: Int,
        val analyzed: Int,
        val errors: List<String>
    )
}
